prospector/filter_rank/ranker.py

```diff
@@ -132,3 +133,57 @@ def make_dataframe(
         )  # This line needs to be updated to correspond to
         # the actual features
     return
+
+
+def apply_rules(
+    candidates: "list[CommitFeatures]", rules="ALL"
+) -> "list[CommitFeatures]":
+    """
+    This applies a set of hand-crafted rules and returns a dict in the following form:
+
+    commits_ruled[candidate] = ["explanation"]
+
+    where 'explanation' describes the rule that matched for that candidate
+    """
+    rule_application_result = dict()
+    for candidate in candidates:
+        # Below goes the code to extract commits that correspond to a particular rule
+        # To add a new rule, one needs to add the following code snippet:
+        # if rules == "ALL" or "Rule name" in rules:
+        #    apply_rule(candidate, rule_application_result)
+        if rules == "ALL" or "REF_VULN_ID" in rules:
+            apply_rule_references_vuln_id(candidate, rule_application_result)
+        if rules == "ALL" or "REF_GH_ISSUE" in rules:
+            apply_rule_references_ghissue(candidate, rule_application_result)
+        if rules == "ALL" or "CH_REL_PATH" in rules:
+            apply_rule_changes_relevant_path(candidate, rule_application_result)
+    # NOTE: the CommitFeatures object has a handy member variable "commit"
+    # which gives access to the underlying "raw" commit object
+    return rule_application_result
+
+
+def apply_rule_references_vuln_id(
+    candidate: CommitFeatures, rule_application_result: dict
+):
+    if candidate.references_vuln_id:
+        if candidate not in rule_application_result:
+            rule_application_result[candidate] = []
+        rule_application_result[candidate].append("Vuln ID is mentioned")
```

Comment:
I would prefer an explicit return statement, rather than modifying a parameter value as a way to return data...
Or am I missing some other reason for adopting this idiom?